doc/designs/quic-design/rx-depacketizer.md

```diff
@@ -0,0 +1,260 @@
+
+```c
+struct ossl_quic_packet_st {
+    SSL *connection;
+
+    /*
+     * |packet_data| is the header directly followed by the payload if there
+     * is any.  |packet_header_length| is the length of the packet header,
+     * and |packet_payload_offset| if the offset to the packet payload if
+     * there is any.
+     *
+     * When passing this structure from the TX packetizer to the QUIC Write
+     * Record Layer, |packet_data| is expected to only hold the packet
+     * header, and |frames| holds the rest of the data.  It's up to the QUIC
+     * Write Record Layer to encrypt and send all those chunks of data in a
+     * manner fitting for the network medium with as little copying as
+     * possible.
+     *
+     * When passing this structure to the RX depacketizer from the QUIC Read
+     * Record Layer, |packet_data| is expected to contain the whole packet,
+     * and |frames| to be NULL.
+     */
+    unsigned char *packet_data;
+    size_t packet_data_length;              /* Only used for RX purposes */
+    size_t packet_header_length;
+    size_t packet_payload_offset;           /* Only used for RX purposes */
+
+    STACK_OF(OSSL_QUIC_FRAME) *frames;      /* Only used for TX purposes */
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+
+    /* Packet types */
+    enum {
+        pkt_initial,
+        pkt_handshake,
+        pkt_rtt0,
+        pkt_rtt1,
+        pkt_retry
+    } packet_type;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+Interactions
+------------
+
+The RX depacketizer receives a packet from the QUIC Read Record Layer, and
+then processes frames in two phases:
+
+1.  [Collect information for the ACK Manager](#collect-information-for-the-ack-manager)
+2.  [Pass frame data](#pass-frame-data)
+
+### Other components
+
+There are a number of other components that the RX depacketizer wants to
+interact with:
+
+-   [ACK manager]
+-   Handshake manager, which is currently unspecified.  It's assumed that
+    this will wrap around what is called the "TLS Handshake Record Layer",
+    in the [overview]
+-   Session manager, which is currently unspecified for QUIC, but may very
+    well be the existing `SSL_SESSION` functionality, extended to fit QUIC
+    purposes.
+-   Flow control, which is currently unspecified.  In the [overview], it's
+    called the "Flow Controller And Statistics Collector"
+-   Connection manager, which is currently unspecified.  In the [overview],
+    there's a "Connection State Machine" that the "RX Frame Handler" isn't
+    talking directly with, so it's possible that the Connection manager will
+    turn out to be the Handshake manager.
+-   Stream SSL objects, to pass the stream data to.
+
+### Receiving or pulling data
+
+The RX depacketizer could be implemented in two ways:
```

Comment:
The design should specify one way I think.
I know it's a bit chicken and egg with lots of the rest undesigned...

Reply:
Thing is that I don't have enough information to base a decision upon.

@mattcaswell and I seem to have different ideas on what's going to drive these bits, but I don't know which one we're ending up with.  Or have we and I didn't notice?

So the question is, is the RX depacketizer supposed to be called from the Read Record Layer (push model) or is it supposed to call the Read Record Layer (pull model)?
Initially, I though that the push model would be used here, but like I said, @mattcaswell seems to have a different idea, but I have still not heard a consensus on the topic.  Until we have a consensus on what the flow of calls is supposed to be, this remains an open question.

Reply:
Seeing #17969, I kinda see @mattcaswell's point...  users of a record layer is supposed to call `read_record`.